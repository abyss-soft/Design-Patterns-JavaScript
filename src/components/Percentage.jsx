import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { getAnswers } from '../selectors';

const StyledPercentage = styled.h1`
  border: 4px solid ${props => props.level};
  border-radius: 50%;
  color: ${props => props.level};
  font-family: 'Karla', 'sans-serif';
  font-size: 4rem;
  height: 10rem;
  margin: auto;
  line-height: 10rem;
  text-align: center;
  width: 10rem;
`;

export const Percentage = ({ answers }) => {
  let level = 'red';
  let correct = 0;

  answers.map(answer => (answer.correct ? correct++ : null));

  const percent = Math.ceil((correct * 100) / 23);

  switch (true) {
  case percent >= 70:
    level = 'limegreen';
    break;
  case percent >= 40:
    level = 'orange';
    break;
  }

  return <StyledPercentage level={level}>{percent}%</StyledPercentage>;
};

Percentage.propTypes = {
  answers: PropTypes.array.isRequired
};

export default connect(state => ({
  answers: getAnswers(state)
}))(Percentage);
