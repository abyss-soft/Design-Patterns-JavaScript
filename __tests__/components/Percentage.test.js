import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { themeCommon } from '../../src/styles/themes/theme.common';
import Percentage from '../../src/components/Percentage';

const mockStore = configureMockStore();
const storeUnder40 = mockStore({
  progress: {
    answers: [{ correct: true }]
  }
});
const storeUnder70 = mockStore({
  progress: {
    answers: [
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true }
    ]
  }
});
const storeOver70 = mockStore({
  progress: {
    answers: [
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: true },
      { correct: false },
      { correct: true }
    ]
  }
});

describe('<Percentage /> component', () => {
  let result;

  beforeEach(() => {
    // the wrapper DIV added to get rid of iterator key warning
    result = shallow(
      <div key="unique">
        <Provider store={storeUnder40}>
          <ThemeProvider theme={themeCommon}>
            <Percentage />
          </ThemeProvider>
        </Provider>
      </div>
    );
  });

  it('renders without crashing', () => {
    expect(result).toBeTruthy();
  });

  it('shows the percentage of correct answers', () => {
    const expected = '5%';
    expect(result.html()).toEqual(expect.stringContaining(expected));
  });

  it('shows the percentage under 40% in red', () => {
    const tree = renderer.create(result).toJSON();
    expect(tree.children[0]).toHaveStyleRule('color', 'red');
  });

  it('shows the percentage under 70% in orange', () => {
    result = shallow(
      <div key="unique">
        <Provider store={storeUnder70}>
          <ThemeProvider theme={themeCommon}>
            <Percentage />
          </ThemeProvider>
        </Provider>
      </div>
    );
    const tree = renderer.create(result).toJSON();
    expect(tree.children[0]).toHaveStyleRule('color', 'orange');
  });

  it('shows the percentage over 70% in green', () => {
    result = shallow(
      <div key="unique">
        <Provider store={storeOver70}>
          <ThemeProvider theme={themeCommon}>
            <Percentage />
          </ThemeProvider>
        </Provider>
      </div>
    );
    const tree = renderer.create(result).toJSON();
    expect(tree.children[0]).toHaveStyleRule('color', 'limegreen');
  });

  it('has styled-component rendered classes', () => {
    const tree = renderer.create(result).toJSON();
    expect(tree.children[0].props.className).toBeDefined();
  });

  it('matches the snapshot', () => {
    const tree = renderer.create(result).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
