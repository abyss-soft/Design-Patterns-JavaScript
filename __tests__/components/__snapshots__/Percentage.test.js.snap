// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Percentage /> component matches the snapshot 1`] = `
.c0 {
  border: 4px solid red;
  border-radius: 50%;
  color: red;
  font-family: 'Karla','sans-serif';
  font-size: 4rem;
  height: 10rem;
  margin: auto;
  line-height: 10rem;
  text-align: center;
  width: 10rem;
}

<div>
  <h1
    className="c0"
  >
    5
    %
  </h1>
</div>
`;
